driver-core/src/main/java/com/datastax/driver/core/SessionManager.java

```diff
@@ -149,12 +149,25 @@ public void run() {
     }
 
     @Override
-    protected ListenableFuture<PreparedStatement> prepareAsync(String query, Map<String, ByteBuffer> customPayload) {
-        Requests.Prepare request = new Requests.Prepare(query);
+    protected ListenableFuture<PreparedStatement> prepareAsync(String query, String keyspace, Map<String, ByteBuffer> customPayload) {
+        try {
+            checkCanSetKeyspace(keyspace);
+        } catch (UnsupportedFeatureException ufe) {
+            return Futures.immediateFailedFuture(ufe);
+        }
+        Requests.Prepare request = new Requests.Prepare(query, keyspace);
         request.setCustomPayload(customPayload);
         Connection.Future future = new Connection.Future(request);
         execute(future, Statement.DEFAULT);
-        return toPreparedStatement(query, future);
+        return toPreparedStatement(query, keyspace, future);
+    }
+
+    private void checkCanSetKeyspace(String keyspace) {
+        if (keyspace != null && !keyspace.equals(poolsState.keyspace) && cluster.manager.protocolVersion().compareTo(ProtocolVersion.V5) < 0) {
```

Comment:
Should I remove the condition that checks if the keyspaces are different?  Maybe we should also throw an exception when using `setKeyspace` at < V5?

Reply:
Just remembered that this is imperfect because it's possible `setKeyspace` was used for something else (like routing).  I think i'll remove this check then

Reply:
Yes I think that's in line with my previous comment about qualified table names.